src/codegen/memmgr.cpp

```diff
@@ -185,15 +207,47 @@ bool PystonMemoryManager::finalizeMemory(std::string* ErrMsg) {
     return false;
 }
 
+static sys::MemoryBlock trimBlockToPageSize(sys::MemoryBlock M) {
+    static const size_t PageSize = sys::Process::getPageSize();
+
+    size_t StartOverlap = (PageSize - ((uintptr_t)M.base() % PageSize)) % PageSize;
+
+    size_t TrimmedSize = M.size();
+    TrimmedSize -= StartOverlap;
+    TrimmedSize -= TrimmedSize % PageSize;
+
+    sys::MemoryBlock Trimmed((void*)((uintptr_t)M.base() + StartOverlap), TrimmedSize);
+
+    assert(((uintptr_t)Trimmed.base() % PageSize) == 0);
+    assert((Trimmed.size() % PageSize) == 0);
+    assert(M.base() <= Trimmed.base() && Trimmed.size() <= M.size());
+
+    return Trimmed;
+}
+
 llvm_error_code PystonMemoryManager::applyMemoryGroupPermissions(MemoryGroup& MemGroup, unsigned Permissions) {
 
-    for (int i = 0, e = MemGroup.AllocatedMem.size(); i != e; ++i) {
+    for (sys::MemoryBlock& MB : MemGroup.PendingMem) {
         llvm_error_code ec;
-        ec = sys::Memory::protectMappedMemory(MemGroup.AllocatedMem[i], Permissions);
+        ec = sys::Memory::protectMappedMemory(MB, Permissions);
         if (ec) {
             return ec;
         }
     }
+    MemGroup.PendingMem.clear();
```

Comment:
@kmod 
According to this [diff file](https://reviews.llvm.org/file/data/vxoazq7dntpccoreh7ro/PHID-FILE-ewe7kqapgt45lwhupue5/D15202.diff)
They might think that they do not have to free pending memory when destructor is called.